Improve memory usage of enrichTags (dogstatsd)

vboulineau opened the pull request:
### What does this PR do?

Small reduction in dogstatsd memory usage in `enrichTags`:

Before:
```
BenchmarkEnrichTags
BenchmarkEnrichTags/20-tags
BenchmarkEnrichTags/20-tags-8            7303458           172 ns/op         224 B/op          1 allocs/op
BenchmarkEnrichTags/40-tags
BenchmarkEnrichTags/40-tags-8            4497426           261 ns/op         416 B/op          1 allocs/op
BenchmarkEnrichTags/60-tags
BenchmarkEnrichTags/60-tags-8            3551148           334 ns/op         640 B/op          1 allocs/op
BenchmarkEnrichTags/80-tags
BenchmarkEnrichTags/80-tags-8            2889858           414 ns/op         896 B/op          1 allocs/op
BenchmarkEnrichTags/100-tags
BenchmarkEnrichTags/100-tags-8           2305017           507 ns/op        1024 B/op          1 allocs/op
BenchmarkEnrichTags/120-tags
BenchmarkEnrichTags/120-tags-8           2100564           580 ns/op        1280 B/op          1 allocs/op
BenchmarkEnrichTags/140-tags
BenchmarkEnrichTags/140-tags-8           1845979           625 ns/op        1408 B/op          1 allocs/op
BenchmarkEnrichTags/160-tags
BenchmarkEnrichTags/160-tags-8           1474875           822 ns/op        1792 B/op          1 allocs/op
BenchmarkEnrichTags/180-tags
BenchmarkEnrichTags/180-tags-8           1470181           802 ns/op        1792 B/op          1 allocs/op
BenchmarkEnrichTags/200-tags
BenchmarkEnrichTags/200-tags-8           1367126           903 ns/op        2048 B/op          1 allocs/op
```

After:
```
BenchmarkEnrichTags/20-tags
BenchmarkEnrichTags/20-tags-8            7106082           176 ns/op         224 B/op          1 allocs/op
BenchmarkEnrichTags/40-tags
BenchmarkEnrichTags/40-tags-8            4521068           255 ns/op         416 B/op          1 allocs/op
BenchmarkEnrichTags/60-tags
BenchmarkEnrichTags/60-tags-8            3717309           331 ns/op         640 B/op          1 allocs/op
BenchmarkEnrichTags/80-tags
BenchmarkEnrichTags/80-tags-8            2713755           420 ns/op         896 B/op          1 allocs/op
BenchmarkEnrichTags/100-tags
BenchmarkEnrichTags/100-tags-8           2346262           517 ns/op        1024 B/op          1 allocs/op
BenchmarkEnrichTags/120-tags
BenchmarkEnrichTags/120-tags-8           2094024           610 ns/op        1280 B/op          1 allocs/op
BenchmarkEnrichTags/140-tags
BenchmarkEnrichTags/140-tags-8           1826889           631 ns/op        1408 B/op          1 allocs/op
BenchmarkEnrichTags/160-tags
BenchmarkEnrichTags/160-tags-8           1490360           811 ns/op        1792 B/op          1 allocs/op
BenchmarkEnrichTags/180-tags
BenchmarkEnrichTags/180-tags-8           1589427           757 ns/op        1792 B/op          1 allocs/op
BenchmarkEnrichTags/200-tags
BenchmarkEnrichTags/200-tags-8           1419412           882 ns/op        2048 B/op          1 allocs/op
```

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Performance optimisation, nothing should change before/after.

vboulineau committed:
Improve memory usage of enrichTags (dogstatsd)

xornivore reviewed (approved):
👍

olivielpeau commented on pkg/dogstatsd/enrich_bench_test.go:
```diff
@@ -0,0 +1,45 @@
+	for i := 100; i < 1000; i *= 2 {
+		b.Run(fmt.Sprintf("%d-tags", i), func(sb *testing.B) {
+			baseTags := buildTags(i)
+			extraTags := buildTags(i / 2)
```
I wonder if the alloc improvement you're seeing in your benchmark isn't simply a result of the dividing value of `2` you've chosen here.
Typically, if the `extraTags` slice always had a length of `1`, I think the results would be worse (since the new code would always allocate a new slice of cap `i+1`, whereas the old code would in most cases not do any allocation).

That said, in most cases, we'd expect the `extraTags` to be longer than `baseTags` right? (when origin tagging is enabled, I'd expect that there are more tags from origin tagging than actual tags on the ingested dogstatsd metrics). If so, then the new code is more efficient indeed. WDYT?

vboulineau replied:
Yes I think you're right, even though I designed the benchmark before changing the code, the benchmark is not optimal in terms of practical use cases.
I've changed it to be closer to reality and it does not show any difference between previous and current code. I'm guessing that the first append creates a large enough slice if `extraTags` is large enough.

Though, with the change you suggested, I think we are now guaranteed to have the minimum number of allocations in any situation, while before it was depending on the size of the different slices.

olivielpeau reviewed (approved):
I don't see anything wrong here, so feel free to merge 👍  I only left a comment for which I'd be interested in your thoughts.

olivielpeau commented on pkg/dogstatsd/enrich.go:
```diff
@@ -40,26 +41,36 @@ func enrichTags(tags []string, defaultHostname string, originTagsFunc func() []s
+		return tags, host
+	}
+
+	finalTags := make([]string, 0, len(tags)+len(originTags)+len(entityTags))
+	finalTags = append(finalTags, tags...)
+	finalTags = append(finalTags, originTags...)
+	finalTags = append(finalTags, entityTags...)
```
actually, you could do this only if `cap(tags) < len(tags)+len(originTags)+len(entityTags)`, otherwise just append to existing tags slice

vboulineau replied:
Good idea, implemented

olivielpeau reviewed (approved):
Can you update the benchmark numbers in the PR description?

I'd also be interested in how the change affects the speed of the function: since the number of allocation hasn't changed, I wonder if the total size allocated matters that much, and if it's visible in the raw speed of the function. Maybe the code changes actually make the function slower for some reason.

In any case, if the updated benchmarks look good, PR LGTM 👍

vboulineau commented:
> Can you update the benchmark numbers in the PR description?
> 
> I'd also be interested in how the change affects the speed of the function: since the number of allocation hasn't changed, I wonder if the total size allocated matters that much, and if it's visible in the raw speed of the function. Maybe the code changes actually make the function slower for some reason.
> 
> In any case, if the updated benchmarks look good, PR LGTM 👍

Updated bench to be in a situation where old code actually needs to re-allocate, updated results in PR, still no diff in the bench though I see with the debugger that `tags` change size twice. Maybe there's some smart reuse of previously allocated blocs, I don't know.
Worst case change does not hurt, best case it brings a little.
